src/main/java/macrobase/analysis/stats/MinCovDet.java

```diff
@@ -148,19 +149,34 @@ private RealVector getMean(List<? extends HasMetrics> data) {
     private List<MetricsWithScore> findKClosest(int k, List<? extends HasMetrics> data) {
         List<MetricsWithScore> scores = new ArrayList<>();
 
+        double[] scoreVec = new double[data.size()];
+
         for (int i = 0; i < data.size(); ++i) {
             HasMetrics d = data.get(i);
-            scores.add(new MetricsWithScore(d.getMetrics(),
-                                            getMahalanobis(mean, inverseCov, d.getMetrics())));
+            Double score = getMahalanobis(mean, inverseCov, d.getMetrics());
+            scores.add(new MetricsWithScore(d.getMetrics(), score));
+            scoreVec[i] = score;
         }
 
         if (scores.size() < k) {
             return scores;
         }
 
-        scores.sort((a, b) -> a.getScore().compareTo(b.getScore()));
+        double cutoff = QuickSelect.select(scoreVec, k, scoreVec.length);
+
+        List<MetricsWithScore> ret = new ArrayList<>(k);
 
-        return scores.subList(0, k);
+        for(MetricsWithScore mws : scores) {
```

Comment:
I see. I saw in the [javadoc](http://ddogleg.org/javadoc/index.html?org/ddogleg/sorting/QuickSelect.html) that there's a ` public static Comparable select(Comparable[] data, int k, int maxIndex, int[] indexes)` interface, where `indexes` is the (output) sorted list of indexes. Maybe you can use this instead of writing your own implementation?